### PR TITLE
add DeliveryModes convenience enum

### PR DESCRIPTION
**Pull request by Jalalx**
## Proposed Changes

This pull request adds the `DeliveryModes` convenience class as a helper to prevent developers from hardcoding delivery mode codes using magic numbers.

## Types of Changes

What types of changes does your code introduce to this project?

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

File: projects/RabbitMQ.Client/client/api/DeliveryModes.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+{
+    /// <summary>
+    /// Convenience class providing compile-time names for persistent modes.
+    /// </summary>
+    public static class DeliveryModes
+    {
+        /// <summary>
+        /// Value for non-persistent delivery mode (not durable).
+        /// </summary>
+        public const byte NonPersistent = 1;
```
Comment (michaelklishin) [inline projects/RabbitMQ.Client/client/api/DeliveryModes.cs:42]: This option is called "transient" in the docs and other clients.

**Comment by michaelklishin** [comment]
Thank you. Besides a bit of work on naming, this looks like a good addition to me.

File: projects/RabbitMQ.Client/client/api/DeliveryModes.cs (hunk 1 above)
Comment (Jalalx) [reply projects/RabbitMQ.Client/client/api/DeliveryModes.cs]: @michaelklishin Thank you for quick review, I just updated the name from `NonPersistent` to `Transient`

**Comment by Matthew-Davey** [comment]
How about an enum?

```csharp
public enum DeliveryMode : byte
{
    Transient = 1,
    Persistent = 2
}
```

**Comment by michaelklishin** [comment]
I like @Matthew-Davey's idea. @Jalalx WDYT about using an enum?

**Comment by Jalalx** [comment]
Using an `enum` was my initial idea, but then two options should be considered:

1. Changing `IBasicProperties.DeliveryMode` from `byte` to `DeliveryModes` enum, which introduces a breaking change to user code and touches too many areas.
2. Leaving `IBasicProperties.DeliveryMode` as is and just add the new `DeliveryModes` enum so user can use it. But then, developer should explicitly convert enum values to `bytes` (personally, I found it annoying 😕).
    ```csharp
    public enum DeliveryModes: byte {  Transient = 1, Persistent = 2 };
    // ...

    IBasicProperties props = ... // 

    // error CS0266: Cannot implicitly convert type 'DeliveryModes' to 'byte'. An explicit conversion exists (are you missing a cast?)
    props.DeliveryMode = DeliveryModes.Transient
    // You have to explicitly convert enum value to byte like this:
    props.DeliveryMode = (byte)DeliveryModes.Transient
    ```
That's why I considered going with simple `cosnt` values in a static `DeliveryModes` class.

Please let me know what you think.

**Comment by bollhals** [comment]
I'd agree with the above statement of @Jalalx .

On the other hand I'm not sure how the usage is divided between the DeliveryMode and the Persistent property. I've personally never used the DeliveryMode. So the benefit for me would be 0.

**Comment by Jalalx** [comment]
Yes, but I wasn't aware of that property until I see that reference while making this PR 😅 I guess `DeliveryMode` is there for backward compatibility because removing it would introduce unnecessary breaking changes to existing client code.

**Comment by sungam3r** [comment]
> Changing IBasicProperties.DeliveryMode from byte to DeliveryModes enum, which introduces a breaking change to user code and touches too many areas.

But it's acceptable for the next major version. See #1056

**Comment by Jalalx** [comment]
@sungam3r Does it worth a breaking change? What do you think @michaelklishin?

**Comment by bollhals** [comment]
> @sungam3r Does it worth a breaking change? What do you think @michaelklishin?

It's not breaking, it's just adding another public class.

**Comment by sungam3r** [comment]
> Does it worth a breaking change?

Why not if v7 is on the way? Nobody likes magic numbers.

**Comment by Jalalx** [comment]
So, should I drop the `DeliveryModes` class and introduce a `DeliveryModes` enum? Is everybody agree on this?

**Comment by lukebakken** [comment]
Thank you @Jalalx

File: projects/RabbitMQ.Client/client/api/BasicProperties.cs [hunk 2]
```diff
@@ -59,8 +59,15 @@ public struct BasicProperties : IBasicProperties, IAmqpHeader
 
         public bool Persistent
         {
-            readonly get { return DeliveryMode == 2; }
-            set { DeliveryMode = value ? (byte)2 : (byte)1; }
+            get
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/BasicProperties.cs:62]: It was just an oversight. Seems like something that could be reviewed through the entire codebase prior to 7.0

**Pull request merged**